### PR TITLE
devex: Add collateral to build the cross-platform NuGet package

### PR DESCRIPTION
**Pull request by HernanGatta**
## Overview

Add instructions and scripts to build the cross-platform NuGet package for Open Enclave. This package carries the combined builds of the SDK for all its supported platforms.

_Current support is for v0.11.0._

### Usage

The Visual Studio extension for Open Enclave helps developers build hosts and enclaves that take advantage of the SDK's multi-TEE support. However, the official NuGet package as released currently only supports SGX on Windows. In contrast, the NuGet package produced via these instructions and scripts includes all the necessary collateral to target SGX on Windows, SGX on Linux, and OP-TEE on Linux.

A number of steps in the build process are currently manual. Ideally, this NuGet package, or a version thereof, would be produced during official builds in a fully automated manner, and would replace the one currently being released.

Signed-off-by: Hernan Gatta <hegatta@microsoft.com>

### CODE REVIEW

File: devex/cross-nuget/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+# Cross-Platform Open Enclave SDK NuGet Package
+
+This directory contains the scripts used to build a cross-platform NuGet package
+for the Open Enclave SDK. The resulting package may be used on Windows to build
```
Comment (dthaler) [inline devex/cross-nuget/README.md:4]: Suggest saying the existence of this package is temporary and issue #3631 tracks unifying this with the other nuget package.

File: devex/cross-nuget/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+   cd /host/path/to/sdk
+   sudo scripts/ansible/install-ansible.sh
+   ansible-playbook scripts/ansible/oe-contributors-setup-cross-arm.yml
+   sudo apt install python3-pyelftools p7zip-full -y
+
+   # Quit the Bionic container.
+   exit
+   ```
+
```
Comment (dthaler) [inline devex/cross-nuget/README.md:78]: Recommend adding a section saying what is needed when bumping the version number.
Are the changes:
1) change the output filename in the instructions below
2) change the <version> value in devex/cross-nuget/extras/open-enclave-cross.nuspec
3) change the OE_SDK_TAG value in devex/cross-nuget/linux/driver.sh
4) Change the version value near the top of devex/cross-nuget/windows/build.ps1

File: devex/cross-nuget/windows/build.ps1 [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# Build for SGX
+# |- Default (with LVI)
+#    |- Debug
+#    |- Release
+
+$ErrorActionPreference = "Stop"
+
+If (-not (Test-Path -Path SDK))
+{
+    git clone --recursive --depth=1 https://github.com/openenclave/openenclave SDK -b v0.11.0
```
Comment (dthaler) [inline devex/cross-nuget/windows/build.ps1:13]: Can "v0.11.0" be put into a powershell variable to something to make it easier to pick out?

File: devex/cross-nuget/README.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+This section explains how to set up both environments.
+
+### Windows Server 2019
+
+1. Install Windows Server 2019 with GUI.
+2. Download the SDK's [Windows prerequisites installation
+   script](https://github.com/openenclave/openenclave/blob/master/scripts/install-windows-prereqs.ps1).
+3. If present, comment out the line in the script that starts with:
+   ```powershell
+   New-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Services\sgx_lc_msr\Parameters" [...]
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:28]: Why is this needed? t.

File: devex/cross-nuget/README.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+   ```
+4. Execute the script as follows:
+   ```powershell
+   Set-ExecutionPolicy -ExecutionPolicy Bypass -Scope Process
+   .\install-windows-prereqs.ps1 -InstallPath C:/oe_prereqs -LaunchConfiguration SGX1FLC-NoIntelDrivers -DCAPClientType None
+   ```
+5. When prompted, reboot the system.
+
+### Ubuntu
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:37]: Can I do this on a VM?

File: devex/cross-nuget/README.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+2. Download the SDK's [Windows prerequisites installation
+   script](https://github.com/openenclave/openenclave/blob/master/scripts/install-windows-prereqs.ps1).
+3. If present, comment out the line in the script that starts with:
+   ```powershell
+   New-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Services\sgx_lc_msr\Parameters" [...]
+   ```
+4. Execute the script as follows:
+   ```powershell
+   Set-ExecutionPolicy -ExecutionPolicy Bypass -Scope Process
+   .\install-windows-prereqs.ps1 -InstallPath C:/oe_prereqs -LaunchConfiguration SGX1FLC-NoIntelDrivers -DCAPClientType None
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:33]: What happens if I already created a VM and installed all the drivers?

File: devex/cross-nuget/README.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+2. Log out and log back in.
+3. Run `setup-containers.sh` to create and start the containers used for
+   building; unlike Docker containers, these remain running in the background
+   and have persistent storage; think of them as virtual machines instead of
+   containers.
+4. Clone the SDK:
+   ```bash
+   git clone -b v0.11.0 --recursive --depth=1 https://github.com/openenclave/openenclave sdk
+   ```
+5. Open `scripts/ansible/oe-contributors-setup.yml` and comment out the step
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:49]: I commented out this
        #    - import_role:
        #name: linux/intel
        #tasks_from: sgx-driver.yml

File: devex/cross-nuget/README.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+   that installs the Intel SGX driver because drivers cannot be installed inside
+   containers.
+6. Install the prerequisites in each container:
+   ```bash
+   # Enter the Xenial container.
+   lxc exec oepkgxenial su ubuntu
+
+   cd /host/path/to/sdk
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:57]: Path

Comment (radhikaj) [reply devex/cross-nuget/README.md:57]: I tried this  $lxc config device add oepkgxenial myhomedir disk source=$HOME path=/dest/ 
But got an error not found. I am sure I am making a mistake, but have no knowledge in this area :)

File: devex/cross-nuget/README.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+## Packaging
+
+Fetch the resulting `pack` directories from both environments and merge their
+contents. This can be achieved by copying the contents of one folder into the
+other; there are no files that overlap.
+
+Then, copy the contents of the `devex/vsextension/msbuild/` directory into the
+`pack/build/native` directory. Additionally, copy the contents of the
+`devex/cross-nuget/extras` directory into the `pack` directory. Update the
+version information insde `open-enclave-cross.nuspec` if necessary.
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:115]: When is it necessary to update the version?

File: devex/cross-nuget/README.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+1. `LICENSE`
+2. `THIRD_PARTY_NOTICES`
+
+To create the NuGet package, navigate to the root of the merged `pack` directory
+and run:
+
+```bash
+# The extension informs 7z as to the type of archive to create.
+7z a -mm=Deflate -mx=9 -mmt=8 -mfb=258 -mpass=15 -r out.zip .
+mv out.zip open-enclave-cross.0.11.0-rc1-cbe4dedc-1.nupkg
```
Comment (radhikaj) [inline devex/cross-nuget/README.md:129]: 0.11.0-rc1-cbe4dedc-1.nupkg
How is cbe4dedc chosen? Is it always the same?
Is 0..11.0.rc1 for v0.11, RC1?

**Comment by radhikaj** [comment]
I got through the steps. I am OK with merging this as long as the versioning questions are answered.

File: devex/cross-nuget/README.md (hunk 1 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md:4]: Updated.

File: devex/cross-nuget/README.md (hunk 2 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md:86]: Thanks for the suggested changes. Added.

File: devex/cross-nuget/windows/build.ps1 (hunk 3 above)
Comment (HernanGatta) [reply devex/cross-nuget/windows/build.ps1]: Indeed, done.

File: devex/cross-nuget/README.md (hunk 4 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md]: I double checked, and it isn't. Removed.

File: devex/cross-nuget/README.md (hunk 6 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md:35]: Nothing. I wrote this under the assumption that the VM used for building does not support SGX. SGX support is not necessary for building, so the drivers don't matter.

File: devex/cross-nuget/README.md (hunk 7 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md:51]: Indeed. I've added a note in the `README` to make that clear.

File: devex/cross-nuget/README.md (hunk 9 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md:135]: Added a section to the `README` with details on versioning. Do you think it's clear now?

File: devex/cross-nuget/README.md (hunk 10 above)
Comment (HernanGatta) [reply devex/cross-nuget/README.md]: Good catch! There was a time when I built a NuGet package for Dave off the tip of `master` instead of a tag, so to indicate which commit the package contained, I included the short commit hash in the name of the package. This is definitely not needed, so I've removed it.

**Comment by HernanGatta** [comment]
> I got through the steps. I am OK with merging this as long as the versioning questions are answered.

Thank you so much for trying them out! I've hopefully addressed your versioning questions.

**Review by radhikaj — APPROVED** [review]

**Review by radhikaj — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: devex/cross-nuget/README.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+Two build systems are required: the first must run Windows Server 2019 with GUI
+and the second must run Ubuntu 18.04 LTS or later (Desktop and Server are both
+OK). These two environments may exist on physical systems or in virtual
+machines, it does not matter.
+
+This section explains how to set up both environments.
+
+### Windows Server 2019
+
+1. Install Windows Server 2019 with GUI.
```
[the remark this answers is not in the document]
Comment (HernanGatta) [reply devex/cross-nuget/README.md:29]: Are you referring to Standard vs Datacenter?

File: devex/cross-nuget/README.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+1. Run `setup-host.sh`, accepting the LXD initialization defaults.
+2. Log out and log back in.
+3. Run `setup-containers.sh` to create and start the containers used for
+   building; unlike Docker containers, these remain running in the background
+   and have persistent storage; think of them as virtual machines instead of
+   containers.
+4. Clone the SDK:
+   ```bash
+   git clone -b v0.11.0 --recursive --depth=1 https://github.com/openenclave/openenclave sdk
```
[the remark this answers is not in the document]
Comment (HernanGatta) [reply devex/cross-nuget/README.md:49]: Is `v0.11.x` a moving target? I would presume that we would build the package for an exact tag for release, and against `master` during CI. Is that not correct?
